Check if updated file exists before adding to list of updated files

dwc0011 opened the pull request:
Fixes #5167 by not adding duplicate files to the updated files list

dwc0011 commented:
I have updated issue to provide a sampel config to illustrate the issue

jeffwidman commented:
Code change looks reasonable to me, with the caveat that I've got minimal terraform expertise.

However, would it be possible to add an rspec test that includes a repo fixture simulating the failure condition of duplicate files? No need to simulate the codecommit response, and I realize GitHub already doesn't fail on this, but it'd still be good to have a test repo to ensure no regressions in the future. Just create a test fixture repo copying your setup from https://github.com/dependabot/dependabot-core/issues/5167#issue-1240553985.

dwc0011 commented:
> Code change looks reasonable to me, with the caveat that I've got minimal terraform expertise.
> 
> However, would it be possible to add an rspec test that includes a repo fixture simulating the failure condition of duplicate files? No need to simulate the codecommit response, and I realize GitHub already doesn't fail on this, but it'd still be good to have a test repo to ensure no regressions in the future. Just create a test fixture repo copying your setup from [#5167 (comment)](https://github.com/dependabot/dependabot-core/issues/5167#issue-1240553985).

I added in a test fixture and updated the rspec, I am still new to ruby and especially the testing stuff so please review, I am sure it will need some expanding on but wanted to see if this was the right start for it.  

Thanks ahead of time.

dwc0011 commented:
After investigating the test further I may need some guidance on how to test it properly.  The terraform file fetcher gathers the files to [process based relative paths etc](https://github.com/dependabot/dependabot-core/blob/50fd5475565065e8e6a1421460b78093e9cc53e8/terraform/lib/dependabot/terraform/file_fetcher.rb#L54)., this is how the duplicate file gets in there.  
But the rspec loads the files differently, by using the [helper](https://github.com/dependabot/dependabot-core/blob/main/bundler/helpers/v2/spec/native_spec_helper.rb) to read all the files in the fixture, this does not duplicate the file.  
I am not sure on the best path for this.  Do we edit the [file fetcher](https://github.com/dependabot/dependabot-core/blob/main/terraform/lib/dependabot/terraform/file_fetcher.rb) instead and prevent duplicate files there or allow the updater to filter out the duplicates to ensure that we do not exclude something prematurely in the file fetcher.  Any thoughts on this would be appreciated.

dwc0011 commented:
Thinking about this more, the current solution gets this working,  maybe use this patch as a quick fix and address the bigger issues as different issue or in one of the related issues such as, #5128 & #5150.
A do think review of how the terraform file fetcher works (processing directories listed in dependabot.yml in addition to resolving relative paths), naming of terraform dependencies for modules and submodules (including git sources, registry and provider sources), and just an overall design review is probably warranted.

Nishnha commented:
Hi! 👋🏾  

Sorry for the silence around this PR--it looks good to me. Thank you for adding appropriate tests.

As for the other linked Terraform issues:
I don't expect this PR to resolve either of them since I think https://github.com/dependabot/dependabot-core/issues/5128 is due to Dependabot not comparing the `variables.tf` file or other similar "metadata" files between versions, and https://github.com/dependabot/dependabot-core/issues/5150 is due to Dependabot having inconsistent in how it treats different Terraform source.

I'll try to take a look at those issues separately sometime this week.

dwc0011 commented:
> Hi! 👋🏾
> 
> Sorry for the silence around this PR--it looks good to me. Thank you for adding appropriate tests.
> 
> As for the other linked Terraform issues: I don't expect this PR to resolve either of them since I think #5128 is due to Dependabot not comparing the `variables.tf` file or other similar "metadata" files between versions, and #5150 is due to Dependabot having inconsistent in how it treats different Terraform source names/versions.
> 
> I'll try to take a look at those issues separately sometime this week.

Nish, thanks for approving this PR.  I didn't mean that this PR would resolve those other PRs rather that those PRs would address the bigger issues of how dependencies are named and files are fetched would fix this issue more appropriately.  
Thanks!
